### DISCUSSION TITLE
Tulip Desktop macOS build questions

### DISCUSSION THREAD

**Discussion by nvillar** [shorepine/tulipcc#144] [General]
Hi,

I'm running into two separate issues building Tulip Desktop on an M2 Mac running Sonoma and Xcode 15 installed.

- When I run ./build.sh it builds successfully, and the app in /dev/ works, except there is no audio (not even the startup beep)
- When I run ./publish.sh, the build fails with the following error:

`
../../micropython/ports/unix/modffi.c:40:10: fatal error: 'ffi.h' file not found. 
#include <ffi.h>
`

I have libffi version 3.4.4 installed via homebrew.

**Comment by bwhitman** [comment]
The first one is a known issue, #103  - you have another audio device on your system and it's trying to play through that, probably like a silent HDMI monitor. It's annoying that our audio engine doesn't choose the same default output as macOS, I'll have to look into that. You can manually specify it with `./tulip -d X` or `./tulip -l` to list them.

I'm not sure I know what `publish.sh` is -- maybe it's `./package.sh`? There's a comment at top about 
```
# Note : on mac you need both types of homebrew, intel and AS
# https://medium.com/mkdir-awesome/how-to-install-x86-64-homebrew-packages-on-apple-m1-macbook-54ba295230f
# arch -x86_64 /usr/local/homebrew/bin/brew install libffi 
```
You need both versions to make a universal binary package. A couple of things (1) i'm not sure we still even need `modffi`, so i'll do some digging there. (2) An average dev / user of Tulip Desktop shouldn't really have to run `package.sh`, that's really just for me to generate the .app package and sign it for distribution as a release on github. `./build.sh` should be enough for everyone!

**Reply by nvillar** [reply]
Thank you for the reply! I'll stick to ./build.sh, I was just curious if the package version would give different results for audio output. The audio in the pre-built Tulip Desktop I downloaded works fine (without having to specify an audio device).

As far as I can tell, I only have 1 audio device. Running tulip with the -l flag gives me a segmentation fault. Running with -d > 0 returns after "No gamepad detected. That's ok" and the window never opens. Running with -d 0 starts as normal, but no audio.

**Reply by bwhitman** [reply]
Oh, weird! If the latest build works but not what you're compiling, that's very interesting. 

I do remember a bug with `tulip -l` crashing, it was a bug in AMY. Can you try removing the file `./.submodules_ok` from the root of the repository and try building again?

**Reply by nvillar** [reply]
I tried removing `./.submodules_ok` and rebuilding, but `tulip -l` still crashes

**Reply by bwhitman** [reply]
I just tried this on a new macOS sonoma install and am getting the same results. I'll dig in !

**Reply by bwhitman** [reply]
I think I fixed it. miniaudio needed a change to support newer macOS . re-pull, and then make sure to remove .submodules_ok again to get the latest AMY. Thank you !

**Reply by nvillar** [reply]
Wonderful, it works! Thank you so much for fixing it. It's great to be able to build Desktop to keep it in sync with Hardware
